File: Sources/UsageWidget/Engine/BackoffPolicy.swift
import Foundation

/// Exponential backoff with jitter, for endpoints that rate-limit aggressively — notably
/// Claude's usage API, which often returns 429 with no `Retry-After`. When the server does
/// provide a delay, we honor it; otherwise we grow the delay geometrically up to a cap.
struct BackoffPolicy: Sendable {
    let base: TimeInterval
    let maxDelay: TimeInterval
    private(set) var attempt: Int = 0

    init(base: TimeInterval = 60, maxDelay: TimeInterval = 30 * 60) {
        self.base = base
        self.maxDelay = maxDelay
    }

    var isBackingOff: Bool { attempt > 0 }

    /// Advance the attempt counter and return the delay before the next try.
    mutating func nextDelay(retryAfter: TimeInterval?) -> TimeInterval {
        attempt += 1
        if let retryAfter, retryAfter > 0 {
            return min(retryAfter, maxDelay)
        }
        let exp = base * pow(2, Double(attempt - 1))
        let capped = min(exp, maxDelay)
        let jitter = Double.random(in: 0...(capped * 0.25))
        return min(capped + jitter, maxDelay)
    }

    mutating func reset() { attempt = 0 }
}

File: windows/src/Core/Engine/BackoffPolicy.cs
namespace LLMUsageWidget.Core.Engine;

/// <summary>Exponential backoff with jitter, for endpoints that rate-limit aggressively — notably
/// Claude's usage API, which often returns 429 with no Retry-After. When the server does provide a
/// delay, we honor it; otherwise we grow the delay geometrically up to a cap. Delays in seconds.</summary>
public sealed class BackoffPolicy
{
    private readonly double _baseSeconds;
    private readonly double _maxSeconds;

    public int Attempt { get; private set; }

    public BackoffPolicy(double baseSeconds = 60, double maxSeconds = 30 * 60)
    {
        _baseSeconds = baseSeconds;
        _maxSeconds = maxSeconds;
    }

    public bool IsBackingOff => Attempt > 0;

    /// <summary>Advance the attempt counter and return the delay before the next try.</summary>
    public double NextDelaySeconds(double? retryAfter)
    {
        Attempt++;
        if (retryAfter is double ra && ra > 0)
            return Math.Min(ra, _maxSeconds);

        double exp = _baseSeconds * Math.Pow(2, Attempt - 1);
        double capped = Math.Min(exp, _maxSeconds);
        double jitter = Random.Shared.NextDouble() * (capped * 0.25);
        return Math.Min(capped + jitter, _maxSeconds);
    }

    public void Reset() => Attempt = 0;
}
